File: fm-domain/src/main/java/io/zeitwert/fm/app/adapter/api/rest/EnumController.java
package io.zeitwert.fm.app.adapter.api.rest;

import dddrive.app.doc.model.enums.CodeCaseStage;
import dddrive.app.doc.model.enums.CodeCaseStageEnum;
import dddrive.ddd.core.model.RepositoryDirectory;
import dddrive.ddd.enums.model.Enumerated;
import dddrive.ddd.enums.model.Enumeration;
import io.zeitwert.dddrive.ddd.api.rest.dto.EnumeratedDto;
import io.zeitwert.fm.app.model.RequestContextFM;
import io.zeitwert.fm.oe.model.ObjTenant;
import io.zeitwert.fm.oe.model.ObjTenantRepository;
import io.zeitwert.fm.oe.model.ObjUser;
import io.zeitwert.fm.oe.model.ObjUserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController("enumController")
@RequestMapping("/enum")
public class EnumController {

	@Autowired
	RequestContextFM requestContext;

	@Autowired
	RepositoryDirectory directory;

	@Autowired
	ObjTenantRepository tenantRepo;

	@Autowired
	ObjUserRepository userRepo;

//	@GetMapping("/oe/objTenant")
//	public ResponseEntity<List<EnumeratedDto>> getTenants() {
//		QuerySpec querySpec = new QuerySpec(ObjUser.class);
//		List<ObjTenantVRecord> tenants = this.tenantRepo.find(querySpec);
//		return ResponseEntity.ok(
//				tenants.stream()
//						.map(obj -> (TableRecord<?>) obj)
//						.map(tenant -> EnumeratedDto.builder().id(tenant.get(ObjFields.ID).toString())
//								.name(tenant.get(ObjFields.CAPTION)).build())
//						.toList());
//	}

	@GetMapping("/oe/objTenant/{id}")
	public ResponseEntity<ObjTenant> getTenant(@PathVariable Integer id) {
		ObjTenant tenant = this.tenantRepo.get(id);
		return ResponseEntity.ok(tenant);
	}

	@GetMapping("/oe/objUser")
	public ResponseEntity<List<EnumeratedDto>> getUsers() {
		Object tenantId = requestContext.getTenantId();
		List<ObjUser> users = userRepo.find(null).stream().map(it -> userRepo.get(it)).toList();
		return ResponseEntity.ok(users.stream().map(u -> EnumeratedDto.of(u.getId().toString(), u.getCaption())).toList());
	}

	@GetMapping("/oe/objUser/{idOrEmail}")
	public ResponseEntity<ObjUser> getUser(@PathVariable String idOrEmail) {
		Optional<ObjUser> user = this.userRepo.getByEmail(idOrEmail);
		if (user.isEmpty()) {
			user = Optional.of(this.userRepo.get(Integer.valueOf(idOrEmail)));
		}
		if (user.isEmpty()) {
			return ResponseEntity.notFound().build();
		}
		return ResponseEntity.ok(user.get());
	}

	@GetMapping("/doc/codeCaseStage/{caseDef}")
	public ResponseEntity<List<CodeCaseStage>> getCaseStageDomain(@PathVariable String caseDef) {
		List<CodeCaseStage> itemList = CodeCaseStageEnum.instance.getItems().stream()
				.filter((item) -> item.getId().startsWith(caseDef + ".") && !"abstract".equals(item.getCaseStageTypeId()))
				.toList();
		return ResponseEntity.ok(itemList);
	}

	@GetMapping("/{module}/{enumerationName}")
	public ResponseEntity<List<? extends Enumerated>> getEnumDomain(@PathVariable String module,
																																	@PathVariable String enumerationName, @RequestParam(required = false, defaultValue = "") List<String> filter) {
		Enumeration<? extends Enumerated> enumeration = directory.getEnumeration(module, enumerationName);
		if (enumeration == null) {
			return ResponseEntity.notFound().build();
		}
		return ResponseEntity.ok().body(enumeration.getItems().stream().filter(i -> !filter.contains(i.getId())).toList());
	}

	@GetMapping("/{module}/{enumerationName}/{id}")
	@SuppressWarnings("unchecked")
	public ResponseEntity<Enumerated> getEnumItem(@PathVariable String module, @PathVariable String enumerationName,
																								@PathVariable String id) {
		Enumerated item = ((Enumeration<Enumerated>) directory.getEnumeration(module, enumerationName)).getItem(id);
		if (item == null) {
			return ResponseEntity.notFound().build();
		}
		return ResponseEntity.ok(item);
	}

}

File: fm-domain/src/main/java/io/zeitwert/app/api/rest/EnumController.kt
package io.zeitwert.app.api.rest

import dddrive.app.doc.model.enums.CodeCaseStageEnum
import dddrive.ddd.model.RepositoryDirectory
import io.zeitwert.app.api.jsonapi.EnumeratedDto
import io.zeitwert.app.session.model.SessionContext
import io.zeitwert.fm.oe.model.ObjTenant
import io.zeitwert.fm.oe.model.ObjTenantRepository
import io.zeitwert.fm.oe.model.ObjUser
import io.zeitwert.fm.oe.model.ObjUserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController("enumController")
@RequestMapping("/enum")
class EnumController {

	@Autowired
	lateinit var sessionContext: SessionContext

	@Autowired
	lateinit var directory: RepositoryDirectory

	@Autowired
	lateinit var tenantRepo: ObjTenantRepository

	@Autowired
	lateinit var userRepo: ObjUserRepository

	// 	@GetMapping("/oe/objTenant")
	// 	public ResponseEntity<List<EnumeratedDto>> getTenants() {
	// 		QuerySpec querySpec = new QuerySpec(ObjUser.class);
	// 		List<ObjTenantVRecord> tenants = tenantRepo.find(querySpec);
	// 		return ResponseEntity.ok(
	// 				tenants.stream()
	// 						.map(obj -> (TableRecord<?>) obj)
	// 						.map(tenant -> EnumeratedDto.builder().id(tenant.get(ObjFields.ID).toString())
	// 								.name(tenant.get(ObjFields.CAPTION)).build())
	// 						.toList());
	// 	}

	@GetMapping("/oe/objTenant/{id}")
	fun getTenant(
		@PathVariable id: Int,
	): ResponseEntity<ObjTenant> {
		val tenant = tenantRepo.get(id)
		return ResponseEntity.ok(tenant)
	}

	@GetMapping("/oe/objUser")
	fun getUsers(): ResponseEntity<List<EnumeratedDto>> {
		sessionContext.tenantId
		val users = userRepo.find(null).map { userRepo.get(it) }
		return ResponseEntity.ok(users.map { EnumeratedDto.of(it.id.toString(), it.caption) })
	}

	@GetMapping("/oe/objUser/{idOrEmail}")
	fun getUser(
		@PathVariable idOrEmail: String,
	): ResponseEntity<ObjUser> {
		try {
			val maybeUser = userRepo.getByEmail(idOrEmail)
			if (maybeUser.isPresent) {
				return ResponseEntity.ok(maybeUser.get())
			}
			val user = userRepo.get(idOrEmail.toInt())
			return ResponseEntity.ok(user)
		} catch (e: Exception) {
			return ResponseEntity.notFound().build()
		}
	}

	@GetMapping("/doc/codeCaseStage/{caseDef}")
	fun getCaseStageDomain(
		@PathVariable caseDef: String,
	): ResponseEntity<List<EnumeratedDto>> {
		val stages = CodeCaseStageEnum.instance.items
			.filter { it.id.startsWith("$caseDef.") && "abstract" != it.caseStageTypeId }
		return ResponseEntity.ok(stages.map { EnumeratedDto.of(it)!! })
	}

	@GetMapping("/{module}/{enumerationName}")
	fun getEnumDomain(
		@PathVariable module: String,
		@PathVariable enumerationName: String,
		@RequestParam(required = false, defaultValue = "") filter: List<String>,
	): ResponseEntity<List<EnumeratedDto>> {
		try {
			val enumeration = directory.getEnumeration(module, enumerationName)
			val items = enumeration.items.filter { !filter.contains(it.id) }.map { EnumeratedDto.of(it)!! }
			return ResponseEntity.ok().body(items)
		} catch (e: Exception) {
			return ResponseEntity.notFound().build()
		}
	}

	@GetMapping("/{module}/{enumerationName}/{id}")
	fun getEnumItem(
		@PathVariable module: String,
		@PathVariable enumerationName: String,
		@PathVariable id: String,
	): ResponseEntity<EnumeratedDto> {
		try {
			val item = directory.getEnumeration(module, enumerationName).getItem(id)
			return ResponseEntity.ok(EnumeratedDto.of(item)!!)
		} catch (e: Exception) {
			return ResponseEntity.notFound().build()
		}
	}

}
